harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +34,107 @@ class DownloadMode(enum.Enum):
     NoSharedDownload = "NO_SHARED_DOWNLOAD"
 
 
+def merge_metadata(all_metadata: List[Dict[str, Any]]) -> Tuple[Dict[str, Any], Dict[str, List]]:
+    """
+      Given a list of metadata, return:
+        - merged metadata
+        - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
+    #  Metadata 1        | Metadata 2          | Conflict |  Result
+    # -------------------|---------------------|----------|-------------------------------
+    #  key1: val         | key2: val           | no       | {key1: val, key2: val}
+    #  key1: val         | key1: val           | yes      | n/a
+    #  key1: [val]       | key1: [val]         | no       | {key1: [val, val]}
+    #  key1: [val]       | key1: val           | yes      | n/a
+    #  key1: [val]       | key1: {key3: val}   | yes      | n/a
+    #  key1: {key3: val} | key1: {key4: val}   | no       | key1: {key3: val, key4: val}
+    #  key1: {key3: val} | key1: val           | yes      | n/a
+    #  key1: {key3: val} | key1: [val]         | yes      | n/a
+    """
+    merged: Dict[str, Any] = {}
+    conflicts: Dict[str, List] = {}
+    for rank, rank_metadata in enumerate(all_metadata):
+
+        for _key in rank_metadata:
+            metadata = rank_metadata[_key]
+            if _key not in merged:
+                merged[_key] = metadata
+                conflicts[_key] = [rank]
+            else:
+                if not isinstance(metadata, type(merged[_key])):
```

Comment:
unrelated to my other comment about this line, you use `merged[key]` quite a few, times in the below code; it might be nice to pull it out as a named value:

```python
for key, new in rank_metadata.items():
    old = merged.get(key)
    if key not in merged:
        merged[key] = new
        conflicts[key] = [rank]
    elif not isinstance(new, type(old)):
        # if values under the same keys have different types
        # report conflict and skip merging
        conflicts[key].append(rank)
    elif isinstance(new, list):
        # merge two lists
        old.extend(new)
    elif isinstance(new, dict):
        # recursively merge dictionaries and collect conflicts
        old, nested_conflicts = merge_metadata([old, new])
        ...
```